File: lesson02/app/src/main/java/br/odb/nehe/lesson02/GL2JNIView.java
package br.odb.nehe.lesson02;

import android.content.Context;
import android.opengl.GLSurfaceView;
import android.util.Log;

import javax.microedition.khronos.egl.EGL10;
import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.egl.EGLContext;
import javax.microedition.khronos.egl.EGLDisplay;
import javax.microedition.khronos.opengles.GL10;

class GL2JNIView extends GLSurfaceView {
    private static final String TAG = "Nehe Lesson02 View";

    public GL2JNIView(Context context) {
        super(context);
        init();
    }

    private void init() {

        setEGLContextFactory(new ContextFactory());

        setEGLConfigChooser(new ConfigChooser(5, 6, 5) );

        setRenderer(new Renderer());
    }

    private static class ContextFactory implements GLSurfaceView.EGLContextFactory {
        private static final int EGL_CONTEXT_CLIENT_VERSION = 0x3098;
        public EGLContext createContext(EGL10 egl, EGLDisplay display, EGLConfig eglConfig) {
            Log.w(TAG, "creating OpenGL ES 2.0 context");
            checkEglError("Before eglCreateContext", egl);
            int[] attrib_list = {EGL_CONTEXT_CLIENT_VERSION, 2, EGL10.EGL_NONE };
            EGLContext context = egl.eglCreateContext(display, eglConfig, EGL10.EGL_NO_CONTEXT, attrib_list);
            checkEglError("After eglCreateContext", egl);
            return context;
        }

        public void destroyContext(EGL10 egl, EGLDisplay display, EGLContext context) {
            egl.eglDestroyContext(display, context);
        }
    }

    private static void checkEglError(String prompt, EGL10 egl) {
        int error;
        while ((error = egl.eglGetError()) != EGL10.EGL_SUCCESS) {
            Log.e(TAG, String.format("%s: EGL error: 0x%x", prompt, error));
        }
    }

    private static class ConfigChooser implements GLSurfaceView.EGLConfigChooser {

        ConfigChooser(int r, int g, int b) {
            mRedSize = r;
            mGreenSize = g;
            mBlueSize = b;
        }
        private static final int EGL_OPENGL_ES2_BIT = 4;
        private static final int[] s_configAttribs2 =
        {
            EGL10.EGL_RED_SIZE, 4,
            EGL10.EGL_GREEN_SIZE, 4,
            EGL10.EGL_BLUE_SIZE, 4,
            EGL10.EGL_RENDERABLE_TYPE, EGL_OPENGL_ES2_BIT,
            EGL10.EGL_NONE
        };

        public EGLConfig chooseConfig(EGL10 egl, EGLDisplay display) {

            int[] num_config = new int[1];
            egl.eglChooseConfig(display, s_configAttribs2, null, 0, num_config);

            int numConfigs = num_config[0];

            if (numConfigs <= 0) {
                throw new IllegalArgumentException("No configs match configSpec");
            }

            EGLConfig[] configs = new EGLConfig[numConfigs];
            egl.eglChooseConfig(display, s_configAttribs2, configs, numConfigs, num_config);

            return chooseConfig(egl, display, configs);
        }

        EGLConfig chooseConfig(EGL10 egl, EGLDisplay display,
                               EGLConfig[] configs) {
            for(EGLConfig config : configs) {

                // We want an *exact* match for red/green/blue/alpha
                int r = findConfigAttrib(egl, display, config,
                        EGL10.EGL_RED_SIZE, 0);
                int g = findConfigAttrib(egl, display, config,
                            EGL10.EGL_GREEN_SIZE, 0);
                int b = findConfigAttrib(egl, display, config,
                            EGL10.EGL_BLUE_SIZE, 0);

                if (r == mRedSize && g == mGreenSize && b == mBlueSize)
                    return config;
            }
            return null;
        }

        private int findConfigAttrib(EGL10 egl, EGLDisplay display,
                EGLConfig config, int attribute, int defaultValue) {

            if (egl.eglGetConfigAttrib(display, config, attribute, mValue)) {
                return mValue[0];
            }
            return defaultValue;
        }

        final int mRedSize;
        final int mGreenSize;
        final int mBlueSize;
        private final int[] mValue = new int[1];
    }

    private static class Renderer implements GLSurfaceView.Renderer {
        public void onDrawFrame(GL10 gl) {
            GL2JNILib.step();
        }

        public void onSurfaceChanged(GL10 gl, int width, int height) {
            GL2JNILib.init(width, height);
        }

        public void onSurfaceCreated(GL10 gl, EGLConfig config) {
            // Do nothing.
        }
    }
}

File: lesson01/app/src/main/java/br/odb/nehe/lesson01/GL2JNIView.kt
package br.odb.nehe.lesson01

import android.content.Context
import android.opengl.GLSurfaceView
import android.util.Log
import br.odb.nehe.lesson01.GL2JNILib.init
import br.odb.nehe.lesson01.GL2JNILib.step
import javax.microedition.khronos.egl.EGL10
import javax.microedition.khronos.egl.EGLConfig
import javax.microedition.khronos.egl.EGLContext
import javax.microedition.khronos.egl.EGLDisplay
import javax.microedition.khronos.opengles.GL10

internal class GL2JNIView(context: Context?) : GLSurfaceView(context) {
	private fun init() {
		setEGLContextFactory(ContextFactory())
		setEGLConfigChooser(ConfigChooser(8, 8, 8))
		setRenderer(Renderer())
	}

	private class ContextFactory : EGLContextFactory {
		override fun createContext(
			egl: EGL10,
			display: EGLDisplay,
			eglConfig: EGLConfig
		): EGLContext {
			Log.w(TAG, "creating OpenGL ES 2.0 context")
			checkEglError("Before eglCreateContext", egl)
			val attrib_list = intArrayOf(EGL_CONTEXT_CLIENT_VERSION, 2, EGL10.EGL_NONE)
			val context =
				egl.eglCreateContext(display, eglConfig, EGL10.EGL_NO_CONTEXT, attrib_list)
			checkEglError("After eglCreateContext", egl)
			return context
		}

		override fun destroyContext(egl: EGL10, display: EGLDisplay, context: EGLContext) {
			egl.eglDestroyContext(display, context)
		}

		companion object {
			private const val EGL_CONTEXT_CLIENT_VERSION = 0x3098
		}
	}

	private class ConfigChooser(
		val mRedSize: Int,
		val mGreenSize: Int,
		val mBlueSize: Int
	) : EGLConfigChooser {
		override fun chooseConfig(egl: EGL10, display: EGLDisplay): EGLConfig {
			val num_config = IntArray(1)
			egl.eglChooseConfig(display, s_configAttribs2, null, 0, num_config)
			val numConfigs = num_config[0]
			require(numConfigs > 0) { "No configs match configSpec" }
			val configs = arrayOfNulls<EGLConfig>(numConfigs)
			egl.eglChooseConfig(display, s_configAttribs2, configs, numConfigs, num_config)
			return chooseConfig(egl, display, configs)!!
		}

		fun chooseConfig(
			egl: EGL10, display: EGLDisplay,
			configs: Array<EGLConfig?>
		): EGLConfig? {
			for (config in configs) {

				// We want an *exact* match for red/green/blue/alpha
				val r = findConfigAttrib(
					egl, display, config,
					EGL10.EGL_RED_SIZE, 0
				)
				val g = findConfigAttrib(
					egl, display, config,
					EGL10.EGL_GREEN_SIZE, 0
				)
				val b = findConfigAttrib(
					egl, display, config,
					EGL10.EGL_BLUE_SIZE, 0
				)
				if (r == mRedSize && g == mGreenSize && b == mBlueSize) return config
			}
			return null
		}

		private fun findConfigAttrib(
			egl: EGL10, display: EGLDisplay,
			config: EGLConfig?, attribute: Int, defaultValue: Int
		): Int {
			return if (egl.eglGetConfigAttrib(display, config, attribute, mValue)) {
				mValue[0]
			} else defaultValue
		}

		private val mValue = IntArray(1)

		companion object {
			private const val EGL_OPENGL_ES2_BIT = 4
			private val s_configAttribs2 = intArrayOf(
				EGL10.EGL_RED_SIZE, 4,
				EGL10.EGL_GREEN_SIZE, 4,
				EGL10.EGL_BLUE_SIZE, 4,
				EGL10.EGL_RENDERABLE_TYPE, EGL_OPENGL_ES2_BIT,
				EGL10.EGL_NONE
			)
		}
	}

	private class Renderer : GLSurfaceView.Renderer {
		override fun onDrawFrame(gl: GL10) {
			step()
		}

		override fun onSurfaceChanged(gl: GL10, width: Int, height: Int) {
			init(width, height)
		}

		override fun onSurfaceCreated(gl: GL10, config: EGLConfig) {
			// Do nothing.
		}
	}

	companion object {
		private const val TAG = "NeHe Lesson01 View"
		private fun checkEglError(prompt: String, egl: EGL10) {
			var error: Int
			while (egl.eglGetError().also { error = it } != EGL10.EGL_SUCCESS) {
				Log.e(TAG, String.format("%s: EGL error: 0x%x", prompt, error))
			}
		}
	}

	init {
		init()
	}
}